File: problems/10_backtracking/0017_letter_combinations_of_a_phone_number/solution.dart
// Problem: 0017. Letter Combinations of a Phone Number
// Difficulty: Medium

List<String> letterCombinations(String digits) {
  if (digits.isEmpty) return <String>[];
  const m = {
    '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
    '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz',
  };
  final out = <String>[];
  final path = StringBuffer();
  final n = digits.length;
  void dfs(int i) {
    if (i == n) {
      out.add(path.toString());
      return;
    }
    final letters = m[digits[i]]!;
    for (var k = 0; k < letters.length; k++) {
      path.write(letters[k]);
      dfs(i + 1);
      final s = path.toString();
      path.clear();
      path.write(s.substring(0, s.length - 1));
    }
  }
  dfs(0);
  return out;
}

bool _eqList(List<String> a, List<String> b) {
  if (a.length != b.length) return false;
  final sa = List<String>.from(a)..sort();
  final sb = List<String>.from(b)..sort();
  for (var i = 0; i < sa.length; i++) if (sa[i] != sb[i]) return false;
  return true;
}

void main() {
  if (!_eqList(letterCombinations("23"),
      ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"])) throw 'a';
  if (letterCombinations("").isNotEmpty) throw 'b';
  if (!_eqList(letterCombinations("2"), ["a", "b", "c"])) throw 'c';
  if (!_eqList(letterCombinations("7"), ["p", "q", "r", "s"])) throw 'd';
  if (letterCombinations("234").length != 27) throw 'e';
  print('All tests passed!');
}

File: problems/10_backtracking/0017_letter_combinations_of_a_phone_number/solution.py
"""
Problem: 0017. Letter Combinations of a Phone Number
Difficulty: Medium
URL: https://leetcode.com/problems/letter-combinations-of-a-phone-number/

解法：回溯。
"""
from __future__ import annotations
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        m = {
            '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
            '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz',
        }
        out: List[str] = []
        path: List[str] = []
        n = len(digits)

        def dfs(i: int) -> None:
            if i == n:
                out.append(''.join(path))
                return
            for ch in m[digits[i]]:
                path.append(ch)
                dfs(i + 1)
                path.pop()

        dfs(0)
        return out


if __name__ == "__main__":
    sol = Solution()
    assert sorted(sol.letterCombinations("23")) == sorted(
        ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"]
    )
    assert sol.letterCombinations("") == []
    assert sol.letterCombinations("2") == ["a", "b", "c"]
    assert sorted(sol.letterCombinations("7")) == ["p", "q", "r", "s"]
    assert len(sol.letterCombinations("234")) == 27
    print("All tests passed!")
